File: programs/kollect/src/instructions/licensing/create_license_template.rs
use anchor_lang::prelude::*;

use crate::constants::{MAX_TEMPLATE_NAME_LENGTH, MAX_URI_LENGTH};
use crate::error::KollectError;
use crate::events::LicenseTemplateCreated;
use crate::state::{LicenseTemplate, TemplateConfig};
use crate::utils::seeds::{LICENSE_TEMPLATE_SEED, TEMPLATE_CONFIG_SEED};

#[derive(Accounts)]
pub struct CreateLicenseTemplate<'info> {
    #[account(mut)]
    pub payer: Signer<'info>,

    #[account(
        mut,
        seeds = [TEMPLATE_CONFIG_SEED],
        bump = template_config.bump,
    )]
    pub template_config: Account<'info, TemplateConfig>,

    #[account(
        init,
        payer = payer,
        space = LicenseTemplate::SIZE,
        seeds = [LICENSE_TEMPLATE_SEED, &template_config.template_count.to_le_bytes()],
        bump,
    )]
    pub license_template: Account<'info, LicenseTemplate>,

    pub system_program: Program<'info, System>,
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone, Debug)]
pub struct CreateLicenseTemplateParams {
    pub template_name: [u8; MAX_TEMPLATE_NAME_LENGTH],
    pub transferable: bool,
    pub derivatives_allowed: bool,
    pub derivatives_reciprocal: bool,
    pub derivatives_approval: bool,
    pub commercial_use: bool,
    pub commercial_attribution: bool,
    pub commercial_rev_share_bps: u16,
    pub derivative_rev_share_bps: u16,
    pub uri: [u8; MAX_URI_LENGTH],
}

pub fn handler(
    ctx: Context<CreateLicenseTemplate>,
    params: CreateLicenseTemplateParams,
) -> Result<()> {
    require!(
        params.commercial_rev_share_bps <= 10_000,
        KollectError::InvalidShareBps
    );
    require!(
        params.derivative_rev_share_bps <= 10_000,
        KollectError::InvalidShareBps
    );

    let clock = Clock::get()?;
    let template_id = ctx.accounts.template_config.template_count;

    // Initialize LicenseTemplate
    let template = &mut ctx.accounts.license_template;
    template.template_id = template_id;
    template.creator = ctx.accounts.payer.key();
    template.template_name = params.template_name;
    template.transferable = params.transferable;
    template.derivatives_allowed = params.derivatives_allowed;
    template.derivatives_reciprocal = params.derivatives_reciprocal;
    template.derivatives_approval = params.derivatives_approval;
    template.commercial_use = params.commercial_use;
    template.commercial_attribution = params.commercial_attribution;
    template.commercial_rev_share_bps = params.commercial_rev_share_bps;
    template.derivative_rev_share_bps = params.derivative_rev_share_bps;
    template.uri = params.uri;
    template.is_active = true;
    template.created_at = clock.unix_timestamp;
    template.bump = ctx.bumps.license_template;

    // Increment counter
    ctx.accounts.template_config.template_count = template_id
        .checked_add(1)
        .ok_or(KollectError::ArithmeticOverflow)?;

    emit!(LicenseTemplateCreated {
        template: template.key(),
        template_id,
        creator: ctx.accounts.payer.key(),
        template_name: params.template_name,
        derivatives_allowed: params.derivatives_allowed,
        commercial_use: params.commercial_use,
        derivative_rev_share_bps: params.derivative_rev_share_bps,
    });

    Ok(())
}

File: scripts/kollect/create_license_template.ts
/**
 * Create License Template Script
 *
 * Usage:
 *   TEMPLATE_NAME="standard_license" URI="ipfs://QmTest" TRANSFERABLE=true \
 *   DERIVATIVES_ALLOWED=true COMMERCIAL_USE=true COMMERCIAL_REV_SHARE_BPS=500 \
 *     anchor run kollect_create_license_template --provider.cluster devnet
 *
 * Environment Variables:
 *   TEMPLATE_NAME              - Template name, max 64 bytes (required)
 *   URI                        - Template URI, max 96 bytes e.g. IPFS CID (required)
 *   TRANSFERABLE               - "true" or "false" (required)
 *   DERIVATIVES_ALLOWED        - "true" or "false" (required)
 *   COMMERCIAL_USE             - "true" or "false" (required)
 *   DERIVATIVES_RECIPROCAL     - "true" or "false" (default: "false")
 *   DERIVATIVES_APPROVAL       - "true" or "false" (default: "false")
 *   COMMERCIAL_ATTRIBUTION     - "true" or "false" (default: "false")
 *   COMMERCIAL_REV_SHARE_BPS   - Basis points 0-10000 (default: 0)
 *   DERIVATIVE_REV_SHARE_BPS   - Basis points 0-10000 (default: 0)
 */

import * as anchor from "@coral-xyz/anchor";
import { Program } from "@coral-xyz/anchor";
import { Kollect } from "../../target/types/kollect";
import {
  deriveTemplateConfigPda,
  deriveLicenseTemplatePda,
  requireEnv,
  getCluster,
  explorerUrl,
  explorerTxUrl,
} from "../../utils/kollect_helper";
import { padBytes } from "../../utils/helper";

const MAX_TEMPLATE_NAME_LENGTH = 64;
const MAX_URI_LENGTH = 96;

function parseBool(value: string, name: string): boolean {
  if (value === "true") return true;
  if (value === "false") return false;
  console.error(`Error: ${name} must be "true" or "false", got "${value}"`);
  process.exit(1);
}

async function main() {
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);

  const program = anchor.workspace.Kollect as Program<Kollect>;
  const authority = provider.wallet;
  const cluster = getCluster(provider.connection.rpcEndpoint);

  console.log("=== Create License Template ===");
  console.log(`Cluster: ${cluster}`);
  console.log(`Payer: ${authority.publicKey.toBase58()}`);
  console.log(`Program ID: ${program.programId.toBase58()}`);

  // --- Parse environment variables ---

  const templateNameStr = requireEnv("TEMPLATE_NAME", "standard_license");
  const uriStr = requireEnv("URI", "ipfs://QmTest");
  const transferableStr = requireEnv("TRANSFERABLE", "true");
  const derivativesAllowedStr = requireEnv("DERIVATIVES_ALLOWED", "true");
  const commercialUseStr = requireEnv("COMMERCIAL_USE", "true");

  const derivativesReciprocalStr =
    process.env.DERIVATIVES_RECIPROCAL ?? "false";
  const derivativesApprovalStr = process.env.DERIVATIVES_APPROVAL ?? "false";
  const commercialAttributionStr =
    process.env.COMMERCIAL_ATTRIBUTION ?? "false";
  const commercialRevShareBpsStr = process.env.COMMERCIAL_REV_SHARE_BPS ?? "0";
  const derivativeRevShareBpsStr = process.env.DERIVATIVE_REV_SHARE_BPS ?? "0";

  // --- Validate ---

  const templateNameBytes = Buffer.from(templateNameStr);
  if (templateNameBytes.length > MAX_TEMPLATE_NAME_LENGTH) {
    console.error(
      `Error: TEMPLATE_NAME exceeds ${MAX_TEMPLATE_NAME_LENGTH} bytes (got ${templateNameBytes.length})`,
    );
    process.exit(1);
  }

  const uriBytes = Buffer.from(uriStr);
  if (uriBytes.length > MAX_URI_LENGTH) {
    console.error(
      `Error: URI exceeds ${MAX_URI_LENGTH} bytes (got ${uriBytes.length})`,
    );
    process.exit(1);
  }

  const transferable = parseBool(transferableStr, "TRANSFERABLE");
  const derivativesAllowed = parseBool(
    derivativesAllowedStr,
    "DERIVATIVES_ALLOWED",
  );
  const derivativesReciprocal = parseBool(
    derivativesReciprocalStr,
    "DERIVATIVES_RECIPROCAL",
  );
  const derivativesApproval = parseBool(
    derivativesApprovalStr,
    "DERIVATIVES_APPROVAL",
  );
  const commercialUse = parseBool(commercialUseStr, "COMMERCIAL_USE");
  const commercialAttribution = parseBool(
    commercialAttributionStr,
    "COMMERCIAL_ATTRIBUTION",
  );

  const commercialRevShareBps = parseInt(commercialRevShareBpsStr, 10);
  const derivativeRevShareBps = parseInt(derivativeRevShareBpsStr, 10);

  if (
    isNaN(commercialRevShareBps) ||
    commercialRevShareBps < 0 ||
    commercialRevShareBps > 10_000
  ) {
    console.error(
      "Error: COMMERCIAL_REV_SHARE_BPS must be between 0 and 10000",
    );
    process.exit(1);
  }
  if (
    isNaN(derivativeRevShareBps) ||
    derivativeRevShareBps < 0 ||
    derivativeRevShareBps > 10_000
  ) {
    console.error(
      "Error: DERIVATIVE_REV_SHARE_BPS must be between 0 and 10000",
    );
    process.exit(1);
  }

  // --- Derive PDAs ---

  const [templateConfigPda] = deriveTemplateConfigPda(program.programId);

  // Fetch template config to get current count
  let templateCount: number;
  try {
    const config = await program.account.templateConfig.fetch(
      templateConfigPda,
    );
    templateCount = (config.templateCount as anchor.BN).toNumber();
  } catch {
    console.error(
      "\nError: TemplateConfig not found. Platform must be initialized first.",
    );
    console.error(
      "Use 'anchor run kollect_initialize_platform' to initialize.",
    );
    process.exit(1);
  }

  const [licenseTemplatePda] = deriveLicenseTemplatePda(
    templateCount,
    program.programId,
  );

  console.log(`\nTemplate Config PDA: ${templateConfigPda.toBase58()}`);
  console.log(`License Template PDA: ${licenseTemplatePda.toBase58()}`);
  console.log(`Next Template ID: ${templateCount}`);
  console.log(`Template Name: ${templateNameStr}`);
  console.log(`URI: ${uriStr}`);
  console.log(`Transferable: ${transferable}`);
  console.log(`Derivatives Allowed: ${derivativesAllowed}`);
  console.log(`Derivatives Reciprocal: ${derivativesReciprocal}`);
  console.log(`Derivatives Approval: ${derivativesApproval}`);
  console.log(`Commercial Use: ${commercialUse}`);
  console.log(`Commercial Attribution: ${commercialAttribution}`);
  console.log(`Commercial Rev Share BPS: ${commercialRevShareBps}`);
  console.log(`Derivative Rev Share BPS: ${derivativeRevShareBps}`);

  // --- Send transaction ---

  console.log("\nCreating license template...");

  try {
    const tx = await program.methods
      .createLicenseTemplate({
        templateName: padBytes(templateNameStr, MAX_TEMPLATE_NAME_LENGTH),
        transferable,
        derivativesAllowed,
        derivativesReciprocal,
        derivativesApproval,
        commercialUse,
        commercialAttribution,
        commercialRevShareBps,
        derivativeRevShareBps,
        uri: padBytes(uriStr, MAX_URI_LENGTH),
      })
      .accountsPartial({
        payer: authority.publicKey,
        templateConfig: templateConfigPda,
        licenseTemplate: licenseTemplatePda,
      })
      .rpc();

    console.log("\n✓ License template created successfully!");
    console.log(`Transaction: ${tx}`);

    // Fetch and display
    const template = await program.account.licenseTemplate.fetch(
      licenseTemplatePda,
    );

    const nameDecoded = Buffer.from(template.templateName as number[])
      .toString("utf8")
      .replace(/\0+$/, "");
    const uriDecoded = Buffer.from(template.uri as number[])
      .toString("utf8")
      .replace(/\0+$/, "");

    console.log("\nTemplate Details:");
    console.log(
      `  Template ID: ${(template.templateId as anchor.BN).toNumber()}`,
    );
    console.log(`  Creator: ${template.creator.toBase58()}`);
    console.log(`  Name: ${nameDecoded}`);
    console.log(`  Transferable: ${template.transferable}`);
    console.log(`  Derivatives Allowed: ${template.derivativesAllowed}`);
    console.log(`  Derivatives Reciprocal: ${template.derivativesReciprocal}`);
    console.log(`  Derivatives Approval: ${template.derivativesApproval}`);
    console.log(`  Commercial Use: ${template.commercialUse}`);
    console.log(`  Commercial Attribution: ${template.commercialAttribution}`);
    console.log(
      `  Commercial Rev Share BPS: ${template.commercialRevShareBps}`,
    );
    console.log(
      `  Derivative Rev Share BPS: ${template.derivativeRevShareBps}`,
    );
    console.log(`  URI: ${uriDecoded}`);
    console.log(`  Is Active: ${template.isActive}`);
    console.log(
      `  Created At: ${new Date(
        (template.createdAt as anchor.BN).toNumber() * 1000,
      ).toISOString()}`,
    );
    console.log(`  Bump: ${template.bump}`);

    console.log(`\nTemplate: ${explorerUrl(licenseTemplatePda, cluster)}`);
    console.log(`Tx: ${explorerTxUrl(tx, cluster)}`);
  } catch (err) {
    console.error("\nError creating license template:", err);
    process.exit(1);
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
